hugo/assets/ts/widgets/carousel.ts

```diff
@@ -0,0 +1,158 @@
+import { BaseWidget } from './base-widget';
+
+export class Carousel extends BaseWidget {
+    public static readonly NAME = 'carousel';
+
+    private readonly slideshow: HTMLUListElement;
+    private readonly amountvisible: number;
+    private readonly dots: NodeListOf<HTMLElement>;
+    private readonly slides: NodeListOf<HTMLElement>;
+    private readonly nextarrow: HTMLElement;
+    private readonly prevarrow: HTMLElement;
+    private readonly scrollLength: number;
+
+    constructor(element: HTMLElement) {
+        super(element);
+
+        this.slideshow = this.element.querySelector<HTMLUListElement>('.carousel__slides');
+        this.amountvisible = Math.round(
+            this.slideshow.offsetWidth /
+            this.slideshow.querySelector<HTMLElement>('.carousel__slide:nth-child(1)').offsetWidth
+        );
+        this.scrollLength = (
+            this.element.querySelector<HTMLElement>('.carousel__slide:nth-child(2)').offsetLeft -
+            this.element.querySelector<HTMLElement>('.carousel__slide:nth-child(1)').offsetLeft
+        );
+        this.dots = this.element.querySelectorAll<HTMLElement>('.carousel__dot');
+        this.slides = this.element.querySelectorAll<HTMLElement>('.carousel__slide');
+        this.nextarrow = this.element.querySelector<HTMLElement>('.carousel__arrow--next');
+        this.prevarrow = this.element.querySelector<HTMLElement>('.carousel__arrow--prev');
+    }
+
+    public static registerWidget(): void {
+        if (window.app !== undefined) {
+            window.app.addWidget({
+                name: Carousel.NAME,
+                load: Carousel.attachWidgetToElements,
+            });
+        }
+    }
+
+    public static attachWidgetToElements(container: HTMLElement | Document): void {
+        const elements = container.querySelectorAll<HTMLElement>(`[data-${ Carousel.NAME }]`);
+        elements.forEach((element) => {
+            const newWidget = new Carousel(element);
+            newWidget.init();
+        });
+    }
+
+    public init(): void {
+        this.slideshow.scrollLeft = 0;
+        this.dots[0].classList.add('selected');
+        this.slides[0].classList.add('selected');
+
+        // Remove dots depending on how many images (== items in shortcode) per slides are shown
+        if (this.amountvisible > 1) {
+            const dotsToRemove = this.element.querySelectorAll(`.carousel__dot:nth-last-child(-n + ${ this.amountvisible - 1 })`);
+            dotsToRemove.forEach((dot) => {
+                dot.remove();
+            });
+        }
+
+        // setInterval for autoplay
+        if (this.element.getAttribute('duration')) {
+        setInterval(() => {
```

Comment:
also save the interval so you can clear it when the component destroys to avoid memory leaks
Maybe also seperate duration & autplay so you're able to set autoplay to true without having to change the duration